### PR TITLE
scheduler for github synchronization

### PR DESCRIPTION
**Pull request by mlachkar**
This pull request adds a scheduler that will run asynchronously to update github infos.
- With a 5000 limit per hour on github api, we are able to synchronize all github info in 8 hours which is quiet slow. There are ways to faster this process: for example only update the readme and repo information and 3 other github request do that less often (once per day for example)
- We used to have 8500 github information, but now we only have +6700. In fact 1200 project are not found anymore. I think it could be nice since all our scaladex relies on github, to have a new field called : GithubStatus (Found, NotFound, Forbiden, MovedPermantly), so we can decide later what to do with project that do not exist anymore.
- Right now, when we cancel githubinfo scheduler, the job is not canceled. this shoul be fixed.

### CODE REVIEW

File: data/src/test/scala/ch.epfl.scala.index/github/LinkParserTests.scala [hunk 1]
```diff
@@ -1,30 +1,30 @@
...
+//package data
+//package github
+//
+//import org.scalatest.funspec.AsyncFunSpec
+//
+//class LinkParserTests extends AsyncFunSpec {
+//  describe("Link parser") {
+//    it("quote") {
+//      assert(
+//        extractLastPage(
```
Comment (adpi2) [inline data/src/test/scala/ch.epfl.scala.index/github/LinkParserTests.scala:11]: Why is this commented?

File: infra/src/main/resources/reference.conf [hunk 2]
```diff
@@ -11,4 +11,8 @@ database {
   name = ${?POSTGRESQL_DATABASE}
   port = 5432
   url = "jdbc:postgresql://"${database.user}":"${database.password}"@localhost:"${database.port}"/"${database.name}
+}
+github {
+  # follow this tutoriel to create your token https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
```
Comment (adpi2) [inline infra/src/main/resources/reference.conf:16]:
```suggestion
  # follow this tutorial to create your token https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
```

File: infra/src/main/scala/ch/epfl/scala/services/github/GithubImplementation.scala [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+import akka.util.ByteString
+import ch.epfl.scala.index.model.misc.GithubInfo
+import ch.epfl.scala.index.model.misc.GithubRepo
+import ch.epfl.scala.index.model.misc.Url
+import ch.epfl.scala.services.GithubService
+import ch.epfl.scala.services.github.GithubModel.Contributor
+import ch.epfl.scala.utils.ScalaExtensions.TraversableOnceFutureExtension
+import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
+
+class GithubImplementation(githubConfig: GithubConfig)(implicit system: ActorSystem) extends GithubService {
```
Comment (adpi2) [inline infra/src/main/scala/ch/epfl/scala/services/github/GithubImplementation.scala:39]: Maybe `AkkaGithubClient` or just `GithubClient`?

File: infra/src/main/scala/ch/epfl/scala/services/github/GithubModel.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      forks_count: Int,
+      mirror_url: Option[String], // "mirror_url": "git://git.apache.org/spark.git",
+      forks: Int,
+      open_issues: Int,
+      default_branch: String, // master
+      subscribers_count: Int, // Watch
+      topics: Seq[String]
+  )
+
+  implicit val repositoryDecoder: Decoder[Repository] = new Decoder[Repository] {
```
Comment (adpi2) [inline infra/src/main/scala/ch/epfl/scala/services/github/GithubModel.scala:32]: Maybe add a comment to explain why the `Decoder` cannot be derived automatically

File: infra/src/main/scala/ch/epfl/scala/services/github/GithubImplementation.scala (hunk 3 above)
Comment (adpi2) [reply infra/src/main/scala/ch/epfl/scala/services/github/GithubImplementation.scala:39]: I don't really like the name `Implementation`...

Comment (mlachkar) [reply infra/src/main/scala/ch/epfl/scala/services/github/GithubImplementation.scala]: Agree

File: infra/src/main/scala/ch/epfl/scala/services/github/GithubModel.scala (hunk 4 above)
Comment (mlachkar) [reply infra/src/main/scala/ch/epfl/scala/services/github/GithubModel.scala:32]: ok

File: data/src/test/scala/ch.epfl.scala.index/github/LinkParserTests.scala (hunk 1 above)
Comment (mlachkar) [reply data/src/test/scala/ch.epfl.scala.index/github/LinkParserTests.scala:11]: I dont know

Comment (mlachkar) [reply data/src/test/scala/ch.epfl.scala.index/github/LinkParserTests.scala:11]: it's the old test of githubReader, that make sure we do the pagination correctly. It's not needed anymore

**Pull request merged**